clang/lib/Serialization/ASTReaderDecl.cpp

```diff
@@ -4099,7 +4099,9 @@ Decl *ASTReader::ReadDeclRecord(DeclID ID) {
   // calls to Decl::getASTContext() by Decl's methods will find the
   // TranslationUnitDecl without crashing.
   D->setDeclContext(Context.getTranslationUnitDecl());
-  Reader.Visit(D);
+
+  // Reading some declarations can result in deep recursion.
+  SemaObj->runWithSufficientStackSpace(DeclLoc, [&] { Reader.Visit(D); });
```

Comment:
Sema may not meaningful in ASTReader (e.g., we're compiling a pcm file). 
```suggestion
clang::runWithSufficientStackSpace(...)
```

Also in this way, we can get better diagnotsics.

Reply:
Could you elaborate in which sense it is non-meaningful for the purpose of calling this method?
I believe SemaObj is never null here (because it's used during deserialization of declarations). The only logic we have in `runWithSufficientStackSpace`, is to warn about stack exhaustion only once and actually show this warning.

I would still be using `Wstack-exhausted` here, not sure there is much useful context to add here (maybe a note about deserializing declarations, but this case is so rare I think we can live without it).

Reply:
As I said, for example, if we're compiling a pcm file to an object file. It makes sense that the Sema is not needed here.

Reply:
I did not realize Sema can be null, I have updated the code as you suggested, but it now duplicates the logic we have in Sema to avoid showing the error multiple times.
Please take a look and see if the new version makes sense.